Bug: test_bgp_update_timer.py fails in L3 scenario due to ExaBGP neighbors flapping

### Is it platform specific

generic

### Importance or Severity

Medium

### Description of the bug

The tests `test_bgp_update_timer_single_route`, `test_bgp_update_timer_session_down` consistently fail in **L3 scenario** (`m0_l3_scenario`) but pass in **VLAN scenario** (`m0_vlan_scenario`).

**Test Setup - L3 scenario**:

```
                    ┌─────────────────────────────────────┐
                    │              DUT                    │
                    │         (BGP AS 65100)              │
                    │                                     │
                    │   20.0.0.0 ◄────────► 20.0.0.2      │
                    └───────┬─────────────────────┬───────┘
                            │                     │
                      IN    │                     │    OUT
                      ▲     │                     │     ▼
              ┌─────────────┴──┐               ┌──┴─────────────┐
              │  n0 (ExaBGP)   │               │  n1 (ExaBGP)   │
              │   20.0.0.1     │               │   20.0.0.3     │
              │  (AS 61000)    │               │  (AS 61001)    │
              └────────────────┘               └────────────────┘
```

**Test Setup - VLAN scenario**:

```
                    ┌─────────────────────────────────────┐
                    │              DUT                    │
                    │         (BGP AS 65100)              │
                    │                                     │
                    │             192.168.0.1             │
                    └───────┬─────────────────────┬───────┘
                            │                     │
                     IN     │                     │    OUT
                      ▲     │                     │     ▼
              ┌─────────────┴──┐               ┌──┴─────────────┐
              │  n0 (ExaBGP)   │               │  n1 (ExaBGP)   │
              │  192.168.0.2   │               │  192.168.0.3   │
              │  (AS 61000)    │               │  (AS 61001)    │
              └────────────────┘               └────────────────┘
```


**Root Cause:**
- In L3 scenario, the test configures secondary IPv4 address (20.0.0.0 and 20.0.0.2) on routed interfaces connected to existing ARISTA BGP neighbors.
- In L3 scenario, test creates ExaBGP neighbors (e.g., 20.0.0.1 and 20.0.0.3) and assigns them to the `PEER_V4` peer-group

IP interface config:
```bash
admin@DUT-1:~$ show ip int
Interface       Master    IPv4 address/mask    Admin/Oper    BGP Neighbor    Neighbor IP
--------------  --------  -------------------  ------------  --------------  -------------
Ethernet46                10.0.0.64/31         up/up         ARISTA01MX      10.0.0.65
Ethernet47                10.0.0.66/31         up/up         ARISTA02MX      10.0.0.67
Loopback0                 10.1.0.32/32         up/up         N/A             N/A
PortChannel101            10.0.0.56/31         up/up         ARISTA01M1      10.0.0.57
                          20.0.0.2/31                        pseudoswitch1   20.0.0.3
PortChannel103            10.0.0.58/31         up/up         ARISTA02M1      10.0.0.59
PortChannel105            10.0.0.60/31         up/up         ARISTA03M1      10.0.0.61
PortChannel106            10.0.0.62/31         up/up         ARISTA04M1      10.0.0.63
                          20.0.0.0/31                        pseudoswitch0   20.0.0.1
Vlan1000                  192.168.0.1/24       up/up         N/A             N/A
docker0                   240.127.1.1/24       up/down       N/A             N/A
eth0                      10.2.139.51/24       up/up         N/A             N/A
lo                        127.0.0.1/16         up/up         N/A             N/A
admin@DUT-1:~$
```

BGP peer groups:
```bash
show runningconfig bgp
...
 neighbor 20.0.0.1 remote-as 61000
 neighbor 20.0.0.1 peer-group PEER_V4
 neighbor 20.0.0.1 description pseudoswitch0
 neighbor 20.0.0.1 timers 3 10
 neighbor 20.0.0.1 timers connect 10
 neighbor 20.0.0.3 remote-as 61001
 neighbor 20.0.0.3 peer-group PEER_V4
 neighbor 20.0.0.3 description pseudoswitch1
 neighbor 20.0.0.3 timers 3 10
 neighbor 20.0.0.3 timers connect 10
```

The `PEER_V4` peer-group has a route-map `TO_BGP_PEER_V4 out` that **permits all routes**:
```bash
route-map TO_BGP_PEER_V4 permit 100
  call CHECK_IDF_ISOLATION
exit
 ...

  neighbor PEER_V4 route-map TO_BGP_PEER_V4 out
```

DUT sends **thousands of routes** learned from ARISTA neighbors to ExaBGP neghbors. ExaBGP cannot handle this flood, causing:
- BGP sessions to flap continuously
- `State/PfxRcd` always flaps between `Idle`, `Active`, or `0` (non-stable neighbors)
- Test assertions to fail because expected BGP UPDATE packets are not captured

**Why L3 scenario fails but VLAN scenario passes:**
In VLAN scenario, ExaBGP neighbors are configured within the `192.168.0.0/24` range on Vlan1000:

```bash
admin@DUT-1:~$ show ip int
Interface       Master    IPv4 address/mask    Admin/Oper    BGP Neighbor    Neighbor IP
--------------  --------  -------------------  ------------  --------------  -------------
Ethernet46                10.0.0.64/31         up/up         ARISTA01MX      10.0.0.65
Ethernet47                10.0.0.66/31         up/up         ARISTA02MX      10.0.0.67
Loopback0                 10.1.0.32/32         up/up         N/A             N/A
PortChannel101            10.0.0.56/31         up/up         ARISTA01M1      10.0.0.57
PortChannel103            10.0.0.58/31         up/up         ARISTA02M1      10.0.0.59
PortChannel105            10.0.0.60/31         up/up         ARISTA03M1      10.0.0.61
PortChannel106            10.0.0.62/31         up/up         ARISTA04M1      10.0.0.63
Vlan1000                  192.168.0.1/24       up/up         pseudoswitch0   192.168.0.2
docker0                   240.127.1.1/24       up/down       N/A             N/A
eth0                      10.2.139.51/24       up/up         N/A             N/A
lo                        127.0.0.1/16         up/up         N/A             N/A
admin@DUT-1:~$
```

**PCAP Evidence:**
- In L3 scenario - captured packets show DUT (20.0.0.0) sending large BGP UPDATE messages with thousands of NLRI entries to ExaBGP (20.0.0.1), causing session instability.
- In VLAN scenario - only desired routes 10.10.100.X/27 announce and withdraw

### Steps to Reproduce

1. Set up a testbed with M0 topology 
2. Run test cases:
	- `bgp/test_bgp_update_timer.py::test_bgp_update_timer_single_route[DUT-1-m0_l3_scenario]`
	- `bgp/test_bgp_update_timer.py::test_bgp_update_timer_session_down[DUT-1-m0_l3_scenario]`
3. Observe test failure with errors like:
```bash
==========================================================
FAILED bgp/test_bgp_update_timer.py::test_bgp_update_timer_single_route[DUT-1-m0_l3_scenario] - Failed: no bgp update withdraw route {'prefix': '10.10.100.64/27', 'nexthop': '20.0.0.1'} from 20.0.0.1 to 20.0.0.0
========================================================================================================================== 1 failed, 1 warning in 870.07s (0:14:30) ==========================================================================================================================
DEBUG:tests.conftest:[log_custom_msg] item: <Function test_bgp_update_timer_single_route[DUT-1-m0_l3_scenario]>
```
```bash
FAILED bgp/test_bgp_update_timer.py::test_bgp_update_timer_single_route[DUT-1-m0_l3_scenario] - Failed: no bgp update announce route {'prefix': '10.10.100.0/27', 'nexthop': '20.0.0.1'} from 20.0.0.2 to 20.0.0.3
========================================================================================================================== 1 failed, 1 warning in 656.74s (0:10:56) ==========================================================================================================================
DEBUG:tests.conftest:[log_custom_msg] item: <Function test_bgp_update_timer_single_route[DUT-1-m0_l3_scenario]>
```

4. During test execution, check BGP neighbor status on DUT. ExaBGP neighbors (20.0.0.1, 20.0.0.3) will show flapping state.
   ```bash
	admin@DUT-1:~$ show ip bgp summary

	IPv4 Unicast Summary:
	BGP router identifier 10.1.0.32, local AS number 65100 vrf-id 0
	BGP table version 5641
	RIB entries 2893, using 370304 bytes of memory
	Peers 8, using 192576 KiB of memory
	Peer groups 4, using 256 bytes of memory


	Neighbhor      V     AS    MsgRcvd    MsgSent    TblVer    InQ    OutQ  Up/Down    State/PfxRcd    NeighborName
	-----------  ---  -----  ---------  ---------  --------  -----  ------  ---------  --------------  --------------
	10.0.0.57      4  65200        858        741      5641      0       0  00:33:56   1439            ARISTA01M1
	10.0.0.59      4  65200        855        741      5641      0       0  00:33:56   1439            ARISTA02M1
	10.0.0.61      4  65200        851        741      5641      0       0  00:33:56   1439            ARISTA03M1
	10.0.0.63      4  65200        854        741      5641      0       0  00:33:56   1439            ARISTA04M1
	10.0.0.65      4  64001        800        741      5641      0       0  00:33:56   2               ARISTA01MX
	10.0.0.67      4  64002        804        742      5641      0       0  00:33:56   2               ARISTA02MX
	20.0.0.1       4  61000        264       1914         0      0       0  00:00:01   Idle            pseudoswitch0
	20.0.0.3       4  61001        532        784      5641      0       0  00:24:53   0               pseudoswitch1

	Total number of neighbors 8
	admin@DUT-1:~$ show ip bgp summary

	IPv4 Unicast Summary:
	BGP router identifier 10.1.0.32, local AS number 65100 vrf-id 0
	BGP table version 5641
	RIB entries 2893, using 370304 bytes of memory
	Peers 8, using 192576 KiB of memory
	Peer groups 4, using 256 bytes of memory


	Neighbhor      V     AS    MsgRcvd    MsgSent    TblVer    InQ    OutQ  Up/Down    State/PfxRcd    NeighborName
	-----------  ---  -----  ---------  ---------  --------  -----  ------  ---------  --------------  --------------
	10.0.0.57      4  65200        859        742      5641      0       0  00:33:58   1439            ARISTA01M1
	10.0.0.59      4  65200        855        742      5641      0       0  00:33:58   1439            ARISTA02M1
	10.0.0.61      4  65200        851        742      5641      0       0  00:33:58   1439            ARISTA03M1
	10.0.0.63      4  65200        855        742      5641      0       0  00:33:58   1439            ARISTA04M1
	10.0.0.65      4  64001        801        742      5641      0       0  00:33:58   2               ARISTA01MX
	10.0.0.67      4  64002        805        743      5641      0       0  00:33:58   2               ARISTA02MX
	20.0.0.1       4  61000        264       1914         0      0       0  00:00:03   Active          pseudoswitch0
	20.0.0.3       4  61001        533        785      5641      0       0  00:24:55   0               pseudoswitch1

	Total number of neighbors 8
	admin@DUT-1:~$ 
	```
	
5. Run the same tests with VLAN scenario - test pass:
- `bgp/test_bgp_update_timer.py::test_bgp_update_timer_single_route[DUT-1-m0_vlan_scenario]`
- `bgp/test_bgp_update_timer.py::test_bgp_update_timer_session_down[DUT-1-m0_vlan_scenario]`

### Actual Behavior and Expected Behavior

**Actual Behavior:**
- DUT floods ExaBGP with thousands of routes from ARISTA neighbors
- In L3 scenario (shared PortChannel interface), ExaBGP sessions flap
- Test fails because expected BGP UPDATE packets are not captured reliably

**Expected Behavior:**
- ExaBGP neighbors should have a route-map that only permits test routes (10.10.100.0/24 range) outbound
- BGP sessions should remain stable throughout the test
- Test should pass consistently in both L3 and VLAN scenarios

**Workaround:**

Manually applying a restrictive route-map on DUT to ExaBGP neighbors after n0, n1 session establishment - fixes the issue:

```bash
vtysh -c "configure terminal" \
      -c "ip prefix-list PL_TEST_ROUTES seq 10 permit 10.10.100.0/24 le 32" \
      -c "route-map RM_TEST_ONLY permit 10" \
      -c "match ip address prefix-list PL_TEST_ROUTES" \
      -c "route-map RM_TEST_ONLY deny 20" \
      -c "router bgp 65100" \
      -c "address-family ipv4 unicast" \
      -c "neighbor 20.0.0.1 route-map RM_TEST_ONLY out" \
      -c "neighbor 20.0.0.3 route-map RM_TEST_ONLY out"

vtysh -c "clear ip bgp 20.0.0.1 soft out"
vtysh -c "clear ip bgp 20.0.0.3 soft out"
```

After applying this, BGP sessions stabilize and the test passes.
```bash
========================================================================================================================= 1 passed, 1 warning in 2637.83s (0:43:57) ==========================================================================================================================
DEBUG:tests.conftest:[log_custom_msg] item: <Function test_bgp_update_timer_single_route[DUT-1-m0_l3_scenario]>
```


### Relevant log output

```shell
See logs in bug description and reproduce steps
```

### Output of `show version`

```shell
admin@DUT-1:~$ show version
SONiC Software Version: SONiC.202511.0-dirty-20260120.230812
SONiC OS Version: 13
Distribution: Debian 13.3
Kernel: 6.12.41+deb13-sonic-arm64
Build commit: bc48386dd + 7 mrvl-patches
Build date: Tue Jan 20 22:04:55 UTC 2026
Built by: yan@yan-arm-vm

Platform: arm64-marvell_rd98DX35xx_cn9131-r0
HwSKU: rd98DX35xx_cn9131
ASIC: marvell-prestera
ASIC Count: 1
Serial Number: NA
Model Number: NA
Hardware Revision: N/A
Uptime: 11:14:54 up 3 days, 4 min,  2 users,  load average: 1.41, 4.10, 3.04
Date: Mon 02 Feb 2026 11:14:54

Docker images:                                                                                                                                                                                                                                                                               REPOSITORY                       TAG                              IMAGE ID       SIZE
docker-orchagent                 202511.0-dirty-20260120.230812   a4afa1070dcd   453MB
docker-orchagent                 latest                           a4afa1070dcd   453MB
docker-fpm-frr                   202511.0-dirty-20260120.230812   950013016fee   493MB
docker-fpm-frr                   latest                           950013016fee   493MB
docker-macsec                    202511.0-dirty-20260120.230812   0949709fb30d   443MB
docker-macsec                    latest                           0949709fb30d   443MB
docker-nat                       202511.0-dirty-20260120.230812   12f6eee3b11b   449MB
docker-nat                       latest                           12f6eee3b11b   449MB
docker-dhcp-relay                202511.0-dirty-20260120.230812   f99a677a0820   417MB
docker-dhcp-relay                latest                           f99a677a0820   417MB
docker-sflow                     202511.0-dirty-20260120.230812   2b3181130646   442MB
docker-sflow                     latest                           2b3181130646   442MB
docker-teamd                     202511.0-dirty-20260120.230812   75b869b3aaa9   440MB
docker-teamd                     latest                           75b869b3aaa9   440MB
docker-dash-ha                   202511.0-dirty-20260120.230812   df9005b09723   457MB
docker-dash-ha                   latest                           df9005b09723   457MB
docker-sonic-bmp                 202511.0-dirty-20260120.230812   8746bd2e4890   412MB
docker-sonic-bmp                 latest                           8746bd2e4890   412MB
docker-eventd                    202511.0-dirty-20260120.230812   a477e5ea3469   407MB
docker-eventd                    latest                           a477e5ea3469   407MB
docker-syncd-mrvl-prestera-rpc   202511.0-dirty-20260120.230812   a8b5ff0d084e   586MB
docker-syncd-mrvl-prestera-rpc   latest                           a8b5ff0d084e   586MB
docker-platform-monitor          202511.0-dirty-20260120.230812   f8f1a71d39ca   770MB
docker-platform-monitor          latest                           f8f1a71d39ca   770MB
docker-telemetry-watchdog        202511.0-dirty-20260120.230812   55aeb4d1d105   520MB
docker-telemetry-watchdog        latest                           55aeb4d1d105   520MB
docker-sonic-mgmt-framework      202511.0-dirty-20260120.230812   f6ebc1ed625b   461MB
docker-sonic-mgmt-framework      latest                           f6ebc1ed625b   461MB
docker-snmp                      202511.0-dirty-20260120.230812   2f1e1b8153ec   434MB
docker-snmp                      latest                           2f1e1b8153ec   434MB
docker-sysmgr                    202511.0-dirty-20260120.230812   ac0aa3b516cf   420MB
docker-sysmgr                    latest                           ac0aa3b516cf   420MB
docker-bmp-watchdog              202511.0-dirty-20260120.230812   9bf6a5430366   408MB
docker-bmp-watchdog              latest                           9bf6a5430366   408MB
docker-telemetry-sidecar         202511.0-dirty-20260120.230812   e091ad4f8ad4   407MB
docker-telemetry-sidecar         latest                           e091ad4f8ad4   407MB
docker-auditd-watchdog           202511.0-dirty-20260120.230812   3b6e548e2188   410MB
docker-auditd-watchdog           latest                           3b6e548e2188   410MB
docker-gnmi-watchdog             202511.0-dirty-20260120.230812   a941ffa6a680   415MB
docker-gnmi-watchdog             latest                           a941ffa6a680   415MB
docker-sonic-gnmi                202511.0-dirty-20260120.230812   5313aeb01f77   519MB
docker-sonic-gnmi                latest                           5313aeb01f77   519MB
docker-router-advertiser         202511.0-dirty-20260120.230812   6d4ff99538b1   408MB
docker-router-advertiser         latest                           6d4ff99538b1   408MB
docker-database                  202511.0-dirty-20260120.230812   96f857d79baa   414MB
docker-database                  latest                           96f857d79baa   414MB
docker-lldp                      202511.0-dirty-20260120.230812   e3722d145c9d   415MB
docker-lldp                      latest                           e3722d145c9d   415MB
docker-mux                       202511.0-dirty-20260120.230812   eed9a05ea7ac   420MB
docker-mux                       latest                           eed9a05ea7ac   420MB
docker-auditd                    202511.0-dirty-20260120.230812   9e99e3cac447   407MB
docker-auditd                    latest                           9e99e3cac447   407MB

admin@DUT-1:~$
```

### Attach files (if any)

[test_bgp_update_timer_single_route_dump.zip](https://github.com/user-attachments/files/25272806/test_bgp_update_timer_single_route_dump.zip)

--- a/tests/bgp/test_bgp_update_timer.py
+++ b/tests/bgp/test_bgp_update_timer.py
@@ -36,6 +36,11 @@ pytestmark = [
 PEER_COUNT = 2
 BGP_LOG_TMPL = "/tmp/bgp%d.pcap"
 BGP_DOWN_LOG_TMPL = "/tmp/bgp_down.pcap"
+# Route-map and prefix-list names used to restrict outbound routes
+# to ExaBGP neighbors so the DUT does not flood them with the full
+# routing table (see https://github.com/sonic-net/sonic-mgmt/issues/22391)
+TEST_ROUTES_PREFIX_LIST = "PL_BGP_UPDATE_TEST"
+TEST_ROUTES_ROUTE_MAP = "RM_BGP_UPDATE_TEST"
 ANNOUNCED_SUBNETS = [
     "10.10.100.0/27",
     "10.10.100.32/27",
@@ -55,6 +60,68 @@ NEIGHBOR_ASN1 = 61001
 NEIGHBOR_PORT0 = 11000
 NEIGHBOR_PORT1 = 11001
 WAIT_TIMEOUT = 120
+
+
+def _apply_outbound_route_filter(duthost, dut_asn, neighbor_ips, is_v6, namespace=DEFAULT_NAMESPACE):
+    """Apply an outbound route-map to ExaBGP neighbors so the DUT only
+    advertises the test prefixes (10.10.100.0/24 or fc00:1x::/48) instead
+    of the full routing table.  Without this, ExaBGP sessions flap on
+    topologies where the test neighbors share routed interfaces with real
+    BGP peers (e.g. M0 L3 scenario).
+
+    See: https://github.com/sonic-net/sonic-mgmt/issues/22391
+    """
+    if is_v6:
+        prefix_match = "fc00:10::/48 le 64"
+    else:
+        prefix_match = "10.10.100.0/24 le 32"
+
+    vtysh_cmds = [
+        "configure terminal",
+        "ip prefix-list {} seq 10 permit {}".format(TEST_ROUTES_PREFIX_LIST, prefix_match),
+        "route-map {} permit 10".format(TEST_ROUTES_ROUTE_MAP),
+        "match ip address prefix-list {}".format(TEST_ROUTES_PREFIX_LIST),
+        "exit",
+        "route-map {} deny 20".format(TEST_ROUTES_ROUTE_MAP),
+        "exit",
+        "router bgp {}".format(dut_asn),
+        "address-family {} unicast".format("ipv6" if is_v6 else "ipv4"),
+    ]
+    for ip in neighbor_ips:
+        vtysh_cmds.append("neighbor {} route-map {} out".format(ip, TEST_ROUTES_ROUTE_MAP))
+    vtysh_cmds.append("exit")  # exit address-family
+    vtysh_cmds.append("exit")  # exit router bgp
+
+    ns_option = "-n {}".format(namespace) if namespace != DEFAULT_NAMESPACE else ""
+    cmd = "vtysh {} {}".format(ns_option, " ".join("-c '{}'".format(c) for c in vtysh_cmds))
+    duthost.shell(cmd)
+
+    # Soft-reset outbound so the filter takes effect immediately
+    for ip in neighbor_ips:
+        duthost.shell("vtysh {} -c 'clear {} bgp {} soft out'".format(
+            ns_option, "ipv6" if is_v6 else "ip", ip
+        ))
+
+
+def _remove_outbound_route_filter(duthost, dut_asn, neighbor_ips, is_v6, namespace=DEFAULT_NAMESPACE):
+    """Remove the outbound route-map and prefix-list added by
+    _apply_outbound_route_filter."""
+    ns_option = "-n {}".format(namespace) if namespace != DEFAULT_NAMESPACE else ""
+
+    vtysh_cmds = [
+        "configure terminal",
+        "router bgp {}".format(dut_asn),
+        "address-family {} unicast".format("ipv6" if is_v6 else "ipv4"),
+    ]
+    for ip in neighbor_ips:
+        vtysh_cmds.append("no neighbor {} route-map {} out".format(ip, TEST_ROUTES_ROUTE_MAP))
+    vtysh_cmds.append("exit")  # exit address-family
+    vtysh_cmds.append("exit")  # exit router bgp
+    vtysh_cmds.append("no route-map {}".format(TEST_ROUTES_ROUTE_MAP))
+    vtysh_cmds.append("no ip prefix-list {}".format(TEST_ROUTES_PREFIX_LIST))
+
+    cmd = "vtysh {} {}".format(ns_option, " ".join("-c '{}'".format(c) for c in vtysh_cmds))
+    duthost.shell(cmd, module_ignore_errors=True)
 
 
 @pytest.fixture
@@ -332,6 +399,13 @@ def test_bgp_update_timer_single_route(
         ):
             pytest.fail("Could not establish bgp sessions")
 
+        # Restrict outbound routes to ExaBGP neighbors to only allow test
+        # prefixes, preventing the DUT from flooding them with the full
+        # routing table which causes session flapping (issue #22391)
+        _apply_outbound_route_filter(
+            duthost, n0.peer_asn, [n0.ip, n1.ip], is_v6_topo, n0.namespace
+        )
+
         announce_intervals = []
         withdraw_intervals = []
         for i, route in enumerate(constants.routes):
@@ -423,6 +497,9 @@ def test_bgp_update_timer_single_route(
             pytest.fail(err_msg % ("withdraw", constants.update_interval_threshold))
 
     finally:
+        _remove_outbound_route_filter(
+            duthost, n0.peer_asn, [n0.ip, n1.ip], is_v6_topo, n0.namespace
+        )
         n0.stop_session()
         n1.stop_session()
         for route in constants.routes:
@@ -455,6 +532,13 @@ def test_bgp_update_timer_session_down(
             lambda: is_neighbor_sessions_established(duthost, (n0, n1)),
         ):
             pytest.fail("Could not establish bgp sessions")
+
+        # Restrict outbound routes to ExaBGP neighbors to only allow test
+        # prefixes, preventing the DUT from flooding them with the full
+        # routing table which causes session flapping (issue #22391)
+        _apply_outbound_route_filter(
+            duthost, n0.peer_asn, [n0.ip, n1.ip], is_v6_topo, n0.namespace
+        )
 
         withdraw_intervals = []
         for _, route in enumerate(constants.routes):
@@ -521,6 +605,9 @@ def test_bgp_update_timer_session_down(
                 pytest.fail("withdraw updates route %s not found" % (route))
 
     finally:
+        _remove_outbound_route_filter(
+            duthost, n0.peer_asn, [n0.ip, n1.ip], is_v6_topo, n0.namespace
+        )
         n0.stop_session()
         n1.stop_session()
         for route in constants.routes:

--- a/tests/bgp/test_bgp_update_timer.py
+++ b/tests/bgp/test_bgp_update_timer.py
@@ -64,7 +64,7 @@ WAIT_TIMEOUT = 120
 
 def _apply_outbound_route_filter(duthost, dut_asn, neighbor_ips, is_v6, namespace=DEFAULT_NAMESPACE):
     """Apply an outbound route-map to ExaBGP neighbors so the DUT only
-    advertises the test prefixes (10.10.100.0/24 or fc00:1x::/48) instead
+    advertises the test prefixes (10.10.100.0/24 or fc00:10::/44) instead
     of the full routing table.  Without this, ExaBGP sessions flap on
     topologies where the test neighbors share routed interfaces with real
     BGP peers (e.g. M0 L3 scenario).
@@ -72,7 +72,8 @@ def _apply_outbound_route_filter(duthost, dut_asn, neighbor_ips, is_v6, namespac
     See: https://github.com/sonic-net/sonic-mgmt/issues/22391
     """
     if is_v6:
-        prefix_match = "fc00:10::/48 le 64"
+        # Cover fc00:10::/64 through fc00:14::/64 (and the whole fc00:10::/44 block)
+        prefix_match = "fc00:10::/44 le 64"
     else:
         prefix_match = "10.10.100.0/24 le 32"
 

--- a/tests/bgp/test_bgp_update_timer.py
+++ b/tests/bgp/test_bgp_update_timer.py
@@ -76,11 +76,12 @@ def _apply_outbound_route_filter(duthost, dut_asn, neighbor_ips, is_v6, namespac
     else:
         prefix_match = "10.10.100.0/24 le 32"
 
+    pfx_kw = "ipv6" if is_v6 else "ip"
     vtysh_cmds = [
         "configure terminal",
-        "ip prefix-list {} seq 10 permit {}".format(TEST_ROUTES_PREFIX_LIST, prefix_match),
+        "{} prefix-list {} seq 10 permit {}".format(pfx_kw, TEST_ROUTES_PREFIX_LIST, prefix_match),
         "route-map {} permit 10".format(TEST_ROUTES_ROUTE_MAP),
-        "match ip address prefix-list {}".format(TEST_ROUTES_PREFIX_LIST),
+        "match {} address prefix-list {}".format(pfx_kw, TEST_ROUTES_PREFIX_LIST),
         "exit",
         "route-map {} deny 20".format(TEST_ROUTES_ROUTE_MAP),
         "exit",
@@ -118,7 +119,7 @@ def _remove_outbound_route_filter(duthost, dut_asn, neighbor_ips, is_v6, namespa
     vtysh_cmds.append("exit")  # exit address-family
     vtysh_cmds.append("exit")  # exit router bgp
     vtysh_cmds.append("no route-map {}".format(TEST_ROUTES_ROUTE_MAP))
-    vtysh_cmds.append("no ip prefix-list {}".format(TEST_ROUTES_PREFIX_LIST))
+    vtysh_cmds.append("no {} prefix-list {}".format("ipv6" if is_v6 else "ip", TEST_ROUTES_PREFIX_LIST))
 
     cmd = "vtysh {} {}".format(ns_option, " ".join("-c '{}'".format(c) for c in vtysh_cmds))
     duthost.shell(cmd, module_ignore_errors=True)
@@ -497,11 +498,13 @@ def test_bgp_update_timer_single_route(
             pytest.fail(err_msg % ("withdraw", constants.update_interval_threshold))
 
     finally:
+        # Stop ExaBGP sessions first so they're gone before we re-open the
+        # full-table flood window by removing the outbound filter.
+        n0.stop_session()
+        n1.stop_session()
         _remove_outbound_route_filter(
             duthost, n0.peer_asn, [n0.ip, n1.ip], is_v6_topo, n0.namespace
         )
-        n0.stop_session()
-        n1.stop_session()
         for route in constants.routes:
             duthost.shell("ip route flush %s" % route["prefix"])
 
@@ -605,10 +608,12 @@ def test_bgp_update_timer_session_down(
                 pytest.fail("withdraw updates route %s not found" % (route))
 
     finally:
+        # Stop ExaBGP sessions first so they're gone before we re-open the
+        # full-table flood window by removing the outbound filter.
+        n0.stop_session()
+        n1.stop_session()
         _remove_outbound_route_filter(
             duthost, n0.peer_asn, [n0.ip, n1.ip], is_v6_topo, n0.namespace
         )
-        n0.stop_session()
-        n1.stop_session()
         for route in constants.routes:
             duthost.shell("ip route flush %s" % route["prefix"])

